Add multi select pattern documentation (#1938)
hashicorp/design-system: website/docs/patterns/multi-select-patterns/partials/interaction-concepts/interaction-concepts.md

> When combining a Multi-select pattern with filtering, sorting, pagination, etc, you _will_ encounter confusing scenarios about how selected results should be persisted across different pages and what the user expects when traversing large data sets.

Reviewer: +1 

Also what about instead of "will" -> "may" and instead of "confusing" -> "complex"? The current wording feels a little negative and doesn't assume we know what the consumer finds "confusing".

> These interaction guidelines are _not exhaustive_, but instead should serve as conversation starters between design and engineering for how best to handle the user experience.

Reviewer 2: ```suggestion
These interaction guidelines are not fully comprehensive. They should instead be considered as conversation starters between design and engineering on how best to optimize the user experience.
```

!!! Info

If you run into unique scenarios that these guidelines don’t account for, we would love to hear about them through our normal [support](/about/support) channels.
!!!

## Consistent communication

The most important method to ensure users are interacting with the results they intend to is to clearly communicate what they have selected, or even that a selection of results exists, **at all times**.

This can help to prevent unnecessary deletions of results or making unintended changes if what is currently selected is unclear.

## Persisting the selection

One of the most common questions around Multi-select patterns stems from when and how selected results are persisted when a user performs different actions on a Table. Consider the following scenario:

A user is viewing a data set consisting of 10,000 results that is presented in a Table. In order to make changes to specific results the user may take the following actions:

1. **Filter** the Table to display only a subset of the results.
2. **Sort** the Table by a specific parameter to order the results in a meaningful way.
3. Traverse through multiple pages of the Table through **Pagination**.

> How the selection is persisted across all of these actions can be confusing and open the user up to making unintended changes if what they have selected is persisted (or not persisted) in a logical manner.

Reviewer: Who is it confusing for? Consumers, end-users? It's a bit unclear.

Author: I've reworded this to clarify a bit and focus more on the end-user

> How the selection is persisted across all of these actions can be confusing and open the user up to making unintended changes if what they have selected is persisted (or not persisted) in a logical manner.

Reviewer 3: +1 to what @heatherlarsen mentioned. Maybe adding in an "end-user" would clarify this a bit more. 
```suggestion
How the selection is persisted across all of these actions can be confusing and open the end-user up to making unintended changes if what they have selected is persisted (or not persisted) in a logical manner.
```

### Within a paginated Table

If a user makes a selection on one page of a paginated Table and progresses to the next page, persisting this selection can be communicated by the [selected count](/patterns/multi-select-patterns?tab=guidelines#selected-count) and can be helpful in scenarios where much of the data is similar or overlaps.

<video loop controls width="100%">
    <source
        src="/assets/patterns/multi-select-patterns/multi-select-pagination-interaction.mp4"
        type="video/mp4"
    />

> </video>

Reviewer: The quality of this video isn't very good. Is there anything we can do to improve it?

Reviewer: Do we always loop our videos?

Author: I had to crunch it down a bit because it's longer than most, but I can up the quality. I do generally loop them, but can remove this if necessary.

Reviewer: Let's keep the loop so it's the same as the others. I just hadn't noticed it before.

> </video>

Reviewer: Watching this video also made me realize that it's really easy to miss the selected count increasing as the rows are selected. This makes me wonder if we need to do something that's more obvious or better couple the selection action and this pattern. I'm curious to hear other's opinions here as well.

Reviewer 3: +1 The video adds to the content. Still, I'm curious if focusing more on the interactions, like selecting rows, then moving to a new page, and then the counter reflecting that is possible in a video? Would need to zoom in on each of those interactions and then cut to the next one. 

For example, when recording, capture only the single action. I hope what I'm sharing makes sense, but let me know! 
<img width="604" alt="Screenshot 2024-02-14 at 5 37 57 PM" src="https://github.com/hashicorp/design-system/assets/124841193/6f0411ab-4e07-4825-9041-0d6d4480aab6">

Reviewer: @andgen404's example made me realize that if a user zooms in and selects elements, they may not see the updated count message because it may be out of view. We should run this interaction by @MelSumner to check if this solution is accessible and what alternatives we may need to document.

Author: I've done some confirming with @MelSumner on this, especially as it relates to DOM order, reflow, and assistive technology. I've added a section on Accessibility with an example that users can interact with to communicate the concept, as well as a section on reflow and viewport sizing in the specs.

> ### When filtering Table results

Reviewer: It may be good to communicate that these are conversation starters for product teams and that we don't (and probably won't) have answers for them to these questions.

Reviewer 3: +1

Author: I do reference that at the top of this page, but I've included a reiteration of this point.

Filtering a data set implies removing or hiding results from rendering within a Table; a **reductive** action. In this scenario, performing a reductive action will impact multiple areas of the interface and surface a number of considerations for persisting the selection.

> - In a Multi-select pattern, should the [total count](/patterns/multi-select-patterns?tab=guidlines#total-count) affected to reflect a reduction in the number of visible results?

Reviewer 4: ```suggestion
- In a Multi-select pattern, should the [total count](/patterns/multi-select-patterns?tab=guidlines#total-count) be affected to reflect a reduction in the number of visible results?
```

> When combining a Multi-select pattern with filtering, sorting, or pagination, you may encounter complex scenarios regarding how selected results should be persisted across pages and what the user expects when traversing large data sets.

Reviewer 2: ```suggestion
When combining a Multi-select table with filtering, sorting, or pagination, you may encounter complex scenarios regarding how selected results should be persisted across pages and what the user expects when traversing large data sets.
```

Filtering a data set implies removing or hiding results from rendering within a Table; a **reductive** action. In this scenario, performing a reductive action will impact multiple areas of the interface and surface a number of considerations for persisting the selection.

> - In a Multi-select pattern, should the [total count](/patterns/multi-select-patterns?tab=guidlines#total-count) be affected to reflect a reduction in the number of visible results?

Reviewer 2: ```suggestion
- In a Multi-select pattern, should the [total count](/patterns/multi-select-patterns?tab=guidlines#total-count) change to reflect a reduction in the number of visible results?
```
